File: erlang/p_08.erl
-module(p_08).
-export([compress/1]).

compress(L) ->
    lists:reverse(compress(L, [])).

compress([], Acc) -> Acc;
compress([X], Acc) ->
    [X|Acc];
compress([X, X|Xs], Acc) ->
    compress([X|Xs], Acc);
compress([X, Y|Ys], Acc) when(X =/= Y) ->
    compress([Y|Ys], [X|Acc]).

File: prolog/p-08.pl

compress(Xs, Ys) :-
    compress(Xs, Zs, []), reverse(Ys, Zs).

compress([], Acc, Acc).
compress([X], [X|Acc], Acc).
compress([X,Y|Zs], Res, Acc) :-
    X \= Y, compress([Y|Zs], Res, [X|Acc]).
compress([X,X|Zs], Res, Acc) :-
    compress([X|Zs], Res, Acc).


